test/test_string_map.cpp

```diff
@@ -0,0 +1,1439 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "./allocator_testing_utils.h"
+#include "rcutils/allocator.h"
+#include "rcutils/error_handling.h"
+#include "rcutils/types/string_map.h"
+
+TEST(test_string_map, lifecycle) {
+  auto allocator = rcutils_get_default_allocator();
+  auto failing_allocator = get_failing_allocator();
+  rcutils_ret_t ret;
+
+  // fini a zero initialized string_map
+  {
+    rcutils_string_map_t string_map = rcutils_get_zero_initialized_string_map();
+    ret = rcutils_string_map_fini(&string_map);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+  }
+
+  // init and then fini and then fini again
+  {
+    rcutils_string_map_t string_map = rcutils_get_zero_initialized_string_map();
+    ret = rcutils_string_map_init(&string_map, 10, allocator);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+    ret = rcutils_string_map_fini(&string_map);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+    ret = rcutils_string_map_fini(&string_map);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+  }
+
+  // init and then fini with 0 capacity
+  {
+    rcutils_string_map_t string_map = rcutils_get_zero_initialized_string_map();
+    ret = rcutils_string_map_init(&string_map, 0, allocator);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+    ret = rcutils_string_map_fini(&string_map);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+  }
+
+  // init on non-zero initialized
+  {
+    rcutils_string_map_t string_map;
+    // dirty the memory, otherwise this is flaky (sometimes the junk memory is null)
+    memset(&string_map, 0x7, sizeof(rcutils_string_map_t));
+    ret = rcutils_string_map_init(&string_map, 10, allocator);
+    EXPECT_EQ(RCUTILS_RET_STRING_MAP_ALREADY_INIT, ret) << rcutils_get_error_string_safe();
+    rcutils_reset_error();
+  }
+
+  // double init
+  {
+    rcutils_string_map_t string_map = rcutils_get_zero_initialized_string_map();
+    ret = rcutils_string_map_init(&string_map, 10, allocator);
+    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string_safe();
+    ret = rcutils_string_map_init(&string_map, 10, allocator);
```

Comment:
i see. I just wondered whether the result is different when calling it twice but with different capacity values.